src/net/mqtt_interface.rs

```diff
@@ -0,0 +1,235 @@
+                .unwrap();
+
+        let mut telemetry_topic: String<consts::U128> = String::new();
+        write!(&mut telemetry_topic, "{}/telemetry", prefix).unwrap();
+
+        let mut response_topic: String<consts::U128> = String::new();
+        write!(&mut response_topic, "{}/log", prefix).unwrap();
+
+        Self {
+            mqtt: RefCell::new(mqtt_client),
+            settings: RefCell::new(S::default()),
+            id: String::from(prefix),
+            clock,
+            phy,
+            telemetry_topic,
+            default_response_topic: response_topic,
+            network_was_reset: false,
+            subscribed: false,
+        }
+    }
+
+    /// Update the MQTT interface and service the network
+    ///
+    /// # Returns
+    /// An option containing an action that should be completed as a result of network servicing.
+    pub fn update(&mut self) -> Option<Action> {
+        // First, service the network stack to process any inbound and outbound traffic.
+        let sleep = match self
+            .mqtt
+            .borrow_mut()
+            .network_stack
+            .poll(self.clock.current_ms())
+        {
+            Ok(updated) => !updated,
+            Err(err) => {
+                log::info!("Network error: {:?}", err);
+                false
+            }
+        };
+
+        // If the PHY indicates there's no more ethernet link, reset the DHCP server in the network
+        // stack.
+        if self.phy.poll_link() == false {
+            // Only reset the network stack once per link reconnection. This prevents us from
+            // sending an excessive number of DHCP requests.
+            if !self.network_was_reset {
+                self.network_was_reset = true;
+                self.mqtt.borrow_mut().network_stack.handle_link_reset();
+            }
+        } else {
+            self.network_was_reset = false;
+        }
+
+        let mqtt_connected = match self.mqtt.borrow_mut().is_connected() {
+            Ok(connected) => connected,
+            Err(minimq::Error::Network(
+                smoltcp_nal::NetworkError::NoIpAddress,
+            )) => false,
+            Err(minimq::Error::Network(error)) => {
+                log::info!("Unexpected network error: {:?}", error);
+                false
+            }
+            Err(error) => {
+                log::warn!("Unexpected MQTT error: {:?}", error);
+                false
+            }
+        };
+
+        // If we're no longer subscribed to the settings topic, but we are connected to the broker,
+        // resubscribe.
+        if !self.subscribed && mqtt_connected {
+            let mut settings_topic: String<consts::U128> = String::new();
+            write!(&mut settings_topic, "{}/settings/#", self.id.as_str())
+                .unwrap();
+
+            self.mqtt
+                .borrow_mut()
+                .subscribe(&settings_topic, &[])
+                .unwrap();
+            self.subscribed = true;
```

Comment:
Let's at least state that we don't see and don't process potential subscription failures and why.

Reply:
Added a note here that we log subscription failures but do not act on them currently. I'm of the opinion there's not much more we can do, as there's either a firmware bug or we aren't compatible with the broker. Feel free to comment otherwise.

Reply:
This sounds like we're trying to address a problem that we expect to have in the future. I would say that let's cross that bridge when we come to it - as of right now, we don't publish the active settings, so this is not an issue.

Reply:
Agreed!